Core/src/Navigation/NavigationStreamHelper.cpp

```diff
@@ -0,0 +1,130 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Navigation/NavigationStreamHelper.hpp"
+
+#include "Acts/Surfaces/Surface.hpp"
+
+bool Acts::NavigationStreamHelper::initializeStream(
+    NavigationStream& stream, const GeometryContext& gctx,
+    const NavigationStream::QueryPoint& queryPoint,
+    BoundaryTolerance cTolerance) {
+  // Position and direction from the query point
+  const Vector3& position = queryPoint.position;
+  const Vector3& direction = queryPoint.direction;
+
+  // De-duplicate first (necessary to deal correctly with multiple
+  // intersections) - sort them by surface pointer
+  std::sort(stream.candidates.begin(), stream.candidates.end(),
+            [](const NavigationStream::Candidate& a,
+               const NavigationStream::Candidate& b) {
+              return (&a.surface()) < (&b.surface());
+            });
+  // Remove duplicates on basis of the surface pointer
+  stream.candidates.erase(
+      std::unique(stream.candidates.begin(), stream.candidates.end(),
+                  [](const NavigationStream::Candidate& a,
+                     const NavigationStream::Candidate& b) {
+                    return (&a.surface()) == (&b.surface());
+                  }),
+      stream.candidates.end());
```

Comment:
This I don't quite understand. I thought we want to keep multiple intersections so we can potentially cross a surface twice? Thinking of a tube modeled as a cylinder for example we will encounter the surface twice without switching volume.

Reply:
This is the initialisation call, i.e. you fill 4 cylinders, you get 4 initial candidates.

Then you run the intersections, if you find multiple valid ones per surface, you remember them.
After the intersection run, you join the ones and get all possible intersections.

If you want to re-initialize the same candidates, you need to de-duplicate first and then run the intersection with potential multiplication again.

Reply:
Moved to `NavigationStream` so that it is callable (if needed).